### PR TITLE
[Zen2] Storage layer WriteStateException propagation

### PR DESCRIPTION
**Pull request by andrershov**
Currently, we only log that `WriteStateException` has occurred, in
`GatewayMetaState`. 

This PR goal is to re-throw `WriteStateException` to upper layers.
If `dirty` flag is set to false, we wrap `WriteStateException` in
`UncheckedIOException`, we prefer unchecked exception not to add
explicit `throws` in the multitude of layers.
If `dirty` flag is set to true - the world is broken. And we need to
halt the JVM. Instead of explicit halting in `GatewayMetaState`, we
prefer to throw `IOError`, which will be subsequently handled by
`ElasticsearchUncaughtExceptionHandler` and JVM will be halted. Sadly,
IOError accepts no message argument, that's why we first wrap
`WriteStateException` with `IOException` and then with `IOError`.

This PR also adds tests for `WriteStateException`.

### CODE REVIEW

**Comment by andrershov** [comment]
run gradle build tests

**Comment by andrershov** [comment]
run gradle build tests 1

**Comment by DaveCTurner** [comment]
We discussed this issue as a team today and concluded that we are ok with the node shutting down with an exception in these situations.

File: server/src/main/java/org/elasticsearch/gateway/WriteStateException.java [hunk 1]
```diff
@@ -38,4 +40,15 @@
     public boolean isDirty() {
         return dirty;
     }
+
+    public void rethrowAsErrorOrUncheckedException(String msg) {
+        if (isDirty()) {
+            // IOError is the best thing we have in java library to indicate that serious IO problem has occurred.
+            // IOError will be caught by ElasticsearchUncaughtExceptionHandler and JVM will be halted.
+            // Sadly, it has no constructor that accepts error message, so we first wrap WriteStateException with IOException.
+            throw new IOError(new IOException(msg + ", storage is dirty", this));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/WriteStateException.java]: I think we do not need the message here: the message indicates the calling method but this is also in the stack trace (and the previous log line) and we would know that the storage was dirty if we have an `IOError` caused by a `WriteStateException`.

File: server/src/main/java/org/elasticsearch/gateway/WriteStateException.java [hunk 2]
```diff
@@ -38,4 +40,15 @@
...
+        } else {
+            throw new UncheckedIOException(msg, this);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/WriteStateException.java]: Similarly I think we don't need to add this message.

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 3]
```diff
@@ -208,8 +208,8 @@ public void setCurrentTerm(long currentTerm) {
             metaStateService.writeManifestAndCleanup("current term changed", manifest);
             previousManifest = manifest;
         } catch (WriteStateException e) {
-            logger.warn("Exception occurred when setting current term", e);
-            //TODO re-throw exception
+            logger.error("Failed to set current term to {}", currentTerm);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Could we log the exception and stack trace too? `logger.error(new ParameterizedMessage("Failed to set current term to {}", currentTerm), e);`

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 4]
```diff
@@ -221,8 +221,8 @@ public void setLastAcceptedState(ClusterState clusterState) {
             incrementalWrite = previousClusterState.term() == clusterState.term();
             updateClusterState(clusterState, previousClusterState);
         } catch (WriteStateException e) {
-            logger.warn("Exception occurred when setting last accepted state", e);
-            //TODO re-throw exception
+            logger.error("Failed to set last accepted state with version {}", clusterState.version());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Could we log the exception and stack trace too?

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Asked for a couple of simplifications.

**Comment by andrershov** [comment]
@DaveCTurner I've finished with the changes, could you please take a look?

**Review by DaveCTurner — APPROVED** [review]
LGTM, thanks @andrershov

**Comment by andrershov** [comment]
@DaveCTurner thanks for your reviewing efforts and bringing it up on the team meeting!

**Pull request merged**